### ISSUE TITLE
Allow an external logger

### ISSUE THREAD

**Issue by Crevil** [jonbodner/proteus#17]
The lib uses [logrus](https://github.com/sirupsen/logrus) for logging warnings and debug messages. It would be great if one could provide the needed logging functions when building the dao. Possibly with `logrus` as a fallback. Bundling `logrus` into this lib seems unnecessary to me. The `log` package from the standard library could be used as well and by this removing the dependency.

Is this something we could do? We can of cause provide a PR for it if so.

**Comment by jonbodner** [comment]
Yeah, that's a great idea. I've been playing with a dependency injection pattern for Go for things like logging; I'm curious to see what you've done.

**Comment by Crevil** [comment]
We are only writing small services (micro services) in Go, so after exploring what the possibilities of DI were, we decided to handle it by-hand ie. wire up all dependencies in the main function.

A pattern we have adopted for things like logging is functional options where we make it optional to register the logger for the given package.

```go
event.NewConsumer(event.WithLogger(log))
```

This is not a great mach for the current API so maybe a global log object could be a better way. There is some concerns with concurrency with this solution and there must be some initial wiring between the sub packages of `proteus`.

```go
proteus.WithLogger(log)
err := proteus.Build(dao, proteus.Postgres)
```

Personally I prefer the former one, but this requires an API change that is not really feasible with the variadic `mappers` parameter of `proteus.Build`.

If we consider an API change we could expose a `proteus.NewBuilder()` constructor and make `Build` a method of that. This gives us a central point to construct and wire up the internal logging setup.

What are your thoughts? Can we do something else that makes it simpler and requires a less intrusive change?

**Comment by jonbodner** [comment]
I agree with you about DI; I don't use a tool to do it for me in Go; I wire it by hand in a function called at startup.

I've been using a Logger variable configured at the package level. It can be done either with a wrapper function or directly against a public variable. I'm not terribly concerned about concurrent modification, because this should be called once at startup and never again. If the logger variable is set via a public function, then sync.Once could be used to guarantee this behavior.

I'd want a default logger to be defined so that if the value is never set, then there's still a logger defined (even if it's a do-nothing logger that just swallows all output).

Option A is a lot of work. Off the top of my head, it would require moving all of the functions in Proteus to methods on a type, and defining a new Build function as a factory method that uses the default (do-nothing) logger, along with a WithLogger function that allows the logger to be specified. This has the advantage of being cleaner (no package-level variables), and it could provide different logging levels for different generated DAOs, but I'm not sure it's worth the effort. 

As a first pass, I would prefer the global log object approach first, then after getting some experience with it, moving on to the other approach.

**Comment by Crevil** [comment]
That sounds reasonable. I'll look into making a package level logger that falls back to a noop logger. 
The `sync.Once` wrapper is a no brainer to include that gives a little certainty of concurrent usage.

**Comment by jonbodner** [comment]
This issue is closed by version v0.6.0.  Please re-open if it seems that something else needs to be addressed.
